Readme.md

# user-agent

  User agent string parser for [nodejs](http://nodejs.org).

## Usage

    var userAgent = require('user-agent')
    userAgent.parse('Mozilla/5.0 (Windows; U; Windows NT 5.1; en) AppleWebKit/526.9 (KHTML, like Gecko) Version/4.0dp1 Safari/526.8')
    // => {
        name: 'safari',
        version: '4.0dp1',
        fullName: 'safaro 4.0dp1', 
        os: 'Windows XP',
        full: '... same as above ...'
      }

## Testing

    $ make test

or

    $ node spec/node.js

## License 

(The MIT License) 

Copyright (c) 2010 Ext JS, Inc.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

1.0.3 / 2011-04-06 
==================

  * Added curl and epiphany parser [straps]
  * Added iPhone support [straps]

1.0.1 / 2011-02-02 
==================

  * refactored

0.0.1 / 2010-05-04
==================

  * Initial release


spec/unit/spec.js

describe '.version'
  it 'should be a triplet'
    userAgent.version.should.match(/^\d+\.\d+\.\d+$/)
  end
end

describe '.parse()'
  before
    parse = function(str, name, version, os){
      var agent = userAgent.parse(str)
      agent.should.have_property 'name', name
      agent.should.have_property 'version', version
      agent.should.have_property 'os', os
      agent.should.have_property 'fullName', name + ' ' + version
    }
  end
  
  describe 'Firefox'
    it 'should work'
      parse(
        'Mozilla/5.0 (X11; U; Linux i686; pl-PL; rv:1.9.0.2) Gecko/20121223 Ubuntu/9.25 (jaunty) Firefox/3.8',
        'firefox',
        '3.8',
        'Linux'
        )
      parse(
        'Mozilla/5.0 (Windows; U; Windows NT 6.1; ru; rv:1.9.2b5) Gecko/20091204 Firefox/3.6b5',
        'firefox',
        '3.6b5',
        'Windows 7'
        )
      parse(
        'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.8.1.5) Gecko/20061201 Firefox/2.0.0.5 (Ubuntu-feisty)',
        'firefox',
        '2.0.0.5',
        'Linux'
        )
    end  
  end
  
  describe 'Internet Explorer'
    it 'should work'
      parse(
        'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.2; Trident/4.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0)',
        'msie',
        '8.0',
        'Windows 7'
        )
      parse(
        'Mozilla/4.0 (compatible; MSIE 7.0b; Windows NT 5.2; .NET CLR 1.1.4322; .NET CLR 2.0.50727; InfoPath.2; .NET CLR 3.0.04506.30)',
        'msie',
        '7.0b',
        'Windows 2003'
        )
      parse(
        'Mozilla/4.0 (compatible; MSIE 6.0b; Windows NT 5.1)',
        'msie',
        '6.0b',
        'Windows XP'
        )
      parse(
        'Mozilla/5.0 (Windows; U; MSIE 6.0; Windows NT 5.1; SV1; .NET CLR 2.0.50727)',
        'msie',
        '6.0',
        'Windows XP'
        )
    end
  end
  
  describe 'Konqueror'
    it 'should work'
      parse(
        'Mozilla/5.0 (compatible; Konqueror/4.2; Linux; X11; x86_64) KHTML/4.2.4 (like Gecko) Fedora/4.2.4-2.fc11',
        'konqueror',
        '4.2',
        'Linux'
        )
      parse(
        'Mozilla/5.0 (compatible; Konqueror/3.1-rc6; i686 Linux; 20021105)',
        'konqueror',
        '3.1-rc6',
        'Linux'
        )
    end
  end
  
  describe 'Chrome'
    it 'should work'
      parse(
        'Mozilla/5.0 (X11; U; Linux i686 (x86_64); en-US) AppleWebKit/532.0 (KHTML, like Gecko) Chrome/4.0.202.2 Safari/532.0',
        'chrome',
        '4.0.202.2',
        'Linux'
        )
      parse(
        'Mozilla/5.0 (Windows; U; Windows NT 5.2; en-US) AppleWebKit/525.13 (KHTML, like Gecko) Chrome/0.2.149.27 Safari/525.13',
        'chrome',
        '0.2.149.27',
        'Windows 2003'
        )
      parse(
        'Mozilla/5.0 (Windows; U; Windows NT 6.0; en-US) AppleWebKit/525.13 (KHTML, like Gecko) Chrome/0.2.149.30 Safari/525.13',
        'chrome',
        '0.2.149.30',
        'Windows Vista'
        )
    end
  end
  
  describe 'Opera'
    it 'should work'
      parse(
        'Opera/9.99 (Windows NT 5.1; U; pl) Presto/9.9.9',
        'opera',
        '9.99',
        'Windows XP'
        )
      parse(
        'Mozilla/5.0 (Linux i686 ; U; en; rv:1.8.1) Gecko/20061208 Firefox/2.0.0 Opera 9.70',
        'opera',
        '9.70',
        'Linux'
        )
      parse(
        'Opera/9.64 (X11; Linux i686; U; Linux Mint; it) Presto/2.1.1',
        'opera',
        '9.64',
        'Linux'
        )
    end
  end
  
  describe 'Safari'
    it 'should work'
      parse(
        'Mozilla/5.0 (Windows; U; Windows NT 5.1; en) AppleWebKit/526.9 (KHTML, like Gecko) Version/4.0dp1 Safari/526.8',
        'safari',
        '4.0dp1',
        'Windows XP')
      parse(
        'Mozilla/5.0 (Windows; U; Windows NT 6.0; en-us) AppleWebKit/531.9 (KHTML, like Gecko) Version/4.0.3 Safari/531.9',
        'safari',
        '4.0.3',
        'Windows Vista'
        )
      parse(
        'Mozilla/5.0 (Windows; U; Windows NT 6.1; en-US) AppleWebKit/532+ (KHTML, like Gecko) Version/4.0.2 Safari/530.19.1',
        'safari',
        '4.0.2',
        'Windows 7'
        )
      parse(
        'Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_5_7; en-us) AppleWebKit/531.2+ (KHTML, like Gecko) Version/4.0.1 Safari/530.18',
        'safari',
        '4.0.1',
        'OS X 10.5'
        )
      parse(
        'Mozilla/5.0 (Windows; U; Windows NT 6.0; ru-RU) AppleWebKit/528.16 (KHTML, like Gecko) Version/4.0 Safari/528.16',
        'safari',
        '4.0',
        'Windows Vista'
        )
      parse(
        'Mozilla/5.0 (Windows; U; Windows NT 5.1; cs-CZ) AppleWebKit/525.28.3 (KHTML, like Gecko) Version/3.2.3 Safari/525.29',
        'safari',
        '3.2.3',
        'Windows XP'
        )
      parse(
        'Mozilla/5.0 (iPad; U; CPU OS 3_2 like Mac OS X; en-us) AppleWebKit/531.21.10 (KHTML, like Gecko) Version/4.0.4 Mobile/7B367 Safari/531.21.10',
        'safari',
        '4.0.4',
        'iPad'
        )
      parse(
        'Mozilla/5.0 (iPhone; U; CPU OS 3_1_2 like Mac OS X; en-us) AppleWebKit/531.21.10 (KHTML, like Gecko) Version/4.0 Mobile/7B367 Safari/531.21.10',
        'safari',
        '4.0',
        'iPhone'
        )
    end
  end

  describe 'Epiphany'
    it 'should work'
      parse(
        'Mozilla/5.0 (X11; U; Linux i686; it-it) AppleWebKit/531.2+ (KHTML, like Gecko) Version/5.0 Safari/531.2+ Ubuntu/10.10 () Epiphany/2.30.2',
        'epiphany',
        '2.30.2',
        'Linux'
        )
    end
  end

  describe 'Curl'
    it 'should work'
      parse(
        'curl/7.21.0 (i686-pc-linux-gnu) libcurl/7.21.0 OpenSSL/0.9.8o zlib/1.2.3.4 libidn/1.18',
        'curl',
        '7.21.0',
        'Linux'
        )
    end
  end

end

lib/user-agent.js

/*!
 * user-agent
 * Copyright(c) 2010-2011 TJ Holowaychuk.
 * Authored by TJ Holowaychuk
 * MIT Licensed
 */

/**
 * Library version.
 */

exports.version = '1.0.3'

/**
 * Parse the given user-agent string into an object of usable data.
 *
 * Example:
 *
 *      var userAgent = require('user-agent')
 *      userAgent.parse('Mozilla/5.0 (Windows; U; Windows NT 5.1; en) AppleWebKit/526.9 (KHTML, like Gecko) Version/4.0dp1 Safari/526.8')
 *      // => { name: 'safari', version: '4.0dp1', os: 'Windows XP', full: '... same string as above ...' }
 *
 * @param  {String} str
 * @return {Object}
 * @api public
 */

exports.parse = function(str) {
  var agent = { full: str, name: name(str) };
  agent.version = version(str, agent.name);
  agent.fullName = agent.name + ' ' + agent.version;
  agent.os = os(str);
  return agent;
};

/**
 * Get the browser version based on the given browser name.
 *
 * @param  {String} str
 * @param  {String} name
 * @return {String}
 * @api private
 */

function version(str, name) {
  if (name === 'safari') name = 'version';
  if (name){
	  return new RegExp(name + '[\\/ ]([\\d\\w\\.-]+)', 'i').exec(str) && RegExp.$1 || '';
  }else{
	  var m=str.match(/version[\/ ]([\d\w\.]+)/i);
	  return m && m.length>1 ? m[1] : '';
  }  
}

/**
 * Supported operating systems.
 */

var operatingSystems = {
    'iPad': /ipad/i
  , 'iPhone': /iphone/i
  , 'Windows Vista': /windows nt 6\.0/i
  , 'Windows 7': /windows nt 6\.\d+/i
  , 'Windows 2003': /windows nt 5\.2+/i
  , 'Windows XP': /windows nt 5\.1+/i
  , 'Windows 2000': /windows nt 5\.0+/i
  , 'OS X $1.$2': /os x (\d+)[._](\d+)/i
  , 'Linux': /linux/i
  , 'Googlebot': /googlebot/i
};

var osNames = Object.keys(operatingSystems);

/**
 * Get operating system from the given user-agent string.
 *
 * @param  {String} str
 * @return {String}
 * @api private
 */

function os(str) {
  var captures;
  for (var i = 0, len = osNames.length; i < len; ++i) {
    if (captures = operatingSystems[osNames[i]].exec(str)) {
      return ~osNames[i].indexOf('$1')
        ? osNames[i].replace(/\$(\d+)/g, function(_, n){
          return captures[n]
        }) : osNames[i];
    }
  }
  return '';
}

/**
 * Supported browser names.
 */

var names = [
   'opera'
 , 'konqueror'
 , 'firefox'
 , 'chrome'
 , 'epiphany'
 , 'safari'
 , 'msie'
 , 'curl'
];

/**
 * Get browser name for the given user-agent string.
 *
 * @param  {String} str
 * @return {String}
 * @api private
 */

function name(str) {
  str = str.toLowerCase();
  for (var i = 0, len = names.length; i < len; ++i) {
    if (str.indexOf(names[i]) !== -1) return names[i];
  }
  return '';
}
